Add EIP: Enshrined Proposer Builder Separation

potuz opened the pull request.

potuz committed:
Add EIP for enshrined Proposer Builder Separation

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+This EIP solves a different set of unrelated important problems. 
+
+- An overwhelming majority of beacon block proposers outsource the construction of the execution payload within their blocks to a third party (henceforth called a *builder*). In order to do so, they request the hash tree root (HTR) of a promised execution payload and submit a `SignedBlindedBeaconBlock` to a trusted party that is tasked with replacing the HTR with the full execution payload (recei ...
```
"latter's actions"

barnabemonnot replied:
Name-dropping fair exchange in there might be worth it

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+This EIP solves a different set of unrelated important problems. 
+
+- An overwhelming majority of beacon block proposers outsource the construction of the execution payload within their blocks to a third party (henceforth called a *builder*). In order to do so, they request the hash tree root (HTR) of a promised execution payload and submit a `SignedBlindedBeaconBlock` to a trusted party that is tasked with replacing the HTR with the full execution payload (recei ...
+- Currently, validators have the time between receiving the full beacon block (including an execution payload) and the attesting deadline (4 seconds in Ethereum mainnet) to perform both consensus and execution state transition functions, check blob data availability and evaluate the new head of the blockchain. The remainder of the slot time is spent during less CPU-intense and critical task. By s ...
```
"is spent doing"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- An overwhelming majority of beacon block proposers outsource the construction of the execution payload within their blocks to a third party (henceforth called a *builder*). In order to do so, they request the hash tree root (HTR) of a promised execution payload and submit a `SignedBlindedBeaconBlock` to a trusted party that is tasked with replacing the HTR with the full execution payload (recei ...
+- Currently, validators have the time between receiving the full beacon block (including an execution payload) and the attesting deadline (4 seconds in Ethereum mainnet) to perform both consensus and execution state transition functions, check blob data availability and evaluate the new head of the blockchain. The remainder of the slot time is spent during less CPU-intense and critical task. By s ...
+- By removing the full execution payload size from the consensus block, it allows for faster network propagation on the critical path. 
+- It removes the increased reorg likeliness of including blob transactions in blocks. 
```
Feel like one level of details lower would be good here. Why is this claim true?

potuz replied:
Builder can send the blobs before seeing attestations. Thanks will add this

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- Currently, validators have the time between receiving the full beacon block (including an execution payload) and the attesting deadline (4 seconds in Ethereum mainnet) to perform both consensus and execution state transition functions, check blob data availability and evaluate the new head of the blockchain. The remainder of the slot time is spent during less CPU-intense and critical task. By s ...
+- By removing the full execution payload size from the consensus block, it allows for faster network propagation on the critical path. 
+- It removes the increased reorg likeliness of including blob transactions in blocks. 
+- It prevents validators from missing attestations and strengthens the LMD weight properties of forkchoice in the event that builders produce invalid payloads. 
```
"of the fork choice rule"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- By removing the full execution payload size from the consensus block, it allows for faster network propagation on the critical path. 
+- It removes the increased reorg likeliness of including blob transactions in blocks. 
+- It prevents validators from missing attestations and strengthens the LMD weight properties of forkchoice in the event that builders produce invalid payloads. 
+- It removes the need of using middleware, not maintained by core developers, used by the vast majority of validators and it removes the need from client teams to continue supporting such middleware, improving thus coordination efforts for testing and deploying future hard forks. 
```
"the need to use", but I disagree with this formulation. There is no need-to-use middleware today. Validators can local-build. What it does is remove the need to use (trusted) middleware *in the event that the validator chooses to not self-build*. I am also unclear whether the part about "the need for client teams to continue supporting" is relevant here, this feels like it has to do with the meta of protocol R&D, not with protocol specifications per se (also it might be wrong, if client teams need to continue supporting e.g., builder connections)

potuz replied:
I can remove the sentence, but validators will definitely support the builder API, just that there is no more need for a middle ware. Validators directly contact builders with the builder API. I can either remove the sentence altogether or add an explicit wording for the need if you want to offload building. Either way the claim about clients needing to support and test a middle ware is true regardless

barnabemonnot replied:
So I think they are two separate points, would do two bullets:
- ePBS removes the need to use trusted middleware in order to delegate block construction to a builder.
- ePBS removes the responsibility of testing said middleware from the client teams, improving ...

I am still unclear whether the second bullet is really relevant to a technical specification, and I could see a world where the question whether MEV-Boost should continue to be tested by client teams is debated.

potuz replied:
Lol my original PR had only the first bullet point. The second one was a request :)

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+### Consensus Layer
+
+The full consensus changes can be found the consensus-specs Github repository. They are separated in 
```
"can be found in the consensus-specs GitHub repository"

barnabemonnot replied:
"They are split between:"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- [P2P](TODO-add-link) changes
+- [Honest validator guide](TODO-add-link) changes
+- A new [honest builder](https://github.com/potuz/consensus-specs/blob/345e95504275d6c44b2bc251c50cb230fefb71b5/specs/_features/epbs/builder.md) guide. 
+- [Fork logic](TODO-add-link) changes. 
```
"Fork choice logic"

potuz replied:
that one is actually the fork logic

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- [Beacon Chain](https://github.com/potuz/consensus-specs/blob/345e95504275d6c44b2bc251c50cb230fefb71b5/specs/_features/epbs/beacon-chain.md) changes
+- [Forkchoice](https://github.com/potuz/consensus-specs/blob/345e95504275d6c44b2bc251c50cb230fefb71b5/specs/_features/epbs/fork-choice.md) changes
+- [P2P](TODO-add-link) changes
+- [Honest validator guide](TODO-add-link) changes
```
This and the lines above don't end with a period, while the next two do.

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+| `PAYLOAD_ABSENT` | `uint8(0)` |
+| `PAYLOAD_PRESENT` | `uint8(1)` | 
+| `PAYLOAD_WITHHELD` | `uint8(2)` | 
+| `PAYLOAD_INVALID_STATUS` | `uint8(3)` |
```
Why does this one have `_STATUS` appended and not the others?

potuz replied:
I'm horrible with names and this is a bit late to change, most names will change surely if this is accepted at implementation time, but names are a huge source of bikeshedding that I would want to avoid at this stage. The original reason is that `PAYLOAD_INVALID` already has a meaning in the spec repo. I would rather keep this one and let someone else when I'm not involved deal with name changing.

barnabemonnot replied:
sgtm!

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- A new `process_payload_attestation` is added to `process_operations`, this function validates the payload timeliness attestations broadcast by the PTC members. 
+- `process_execution_payload` is now called as a separate helper when receiving a `SignedExecutionPayloadEnvelope` on the P2P layer. This function in particular checks that the HTR of the resulting beacon state coincides with the committed one in the payload envelope. 
+
+Forkchoice changes include:
```
"Fork choice"

barnabemonnot commented on EIPS/eip-epbs.md:
````diff
@@ -0,0 +1,260 @@
+    root: Root
+```
+
+The class `Store` is modified to track the following fields
````
"fields:"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+### Staked Builders
+
+Being a builder is a new attribution of validators. As such builders are staked in the beacon chain. This allows for in-protocol trustless enforcement of the builder's payment to the proposer. Alternatively, payment could be enforced in the Execution Layer at the cost of adding the corresponding EL consensus changing logic. Payments in the EL have the advantage of not requiring the builder to per ...
```
"Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?

barnabemonnot replied:
"Execution Layer (EL)"

barnabemonnot replied:
"consensus-changing"

barnabemonnot replied:
"in the Consensus Layer (CL)"

barnabemonnot replied:
"must be done"

barnabemonnot replied:
"Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. *How* they are available depends on the layer (EL or CL).

potuz replied:
> "Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?

Aren't these two sentences semantically equivalent?

potuz replied:
> "Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. How they are available depends on the layer (EL or CL).

The point I am making is that the funds have to be available, it doesn't matter the layer. We do not know how to allow for payments with funds that are extracted from the block that the builder is constructed. This construct in principle can be done today with Mev Boost

barnabemonnot replied:
> > "Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?
> 
> Aren't these two sentences semantically equivalent?

Yeah perhaps, I just wasn't 100% clear what the first meant, but apparently I did get it. It's more of a suggestion.

barnabemonnot replied:
> > "Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. How they are available depends on the layer (EL or CL).
> 
> The point I am making is that the funds have to be available, it doesn't matter the layer. We do not know how to allow for payments with funds that are extracted from the block that the builder is constructed. This construct in principle can be done today with Mev Boost

It's just as written it sounds like "because the funds are staked, they must be available". I would just re-write a bit to make clear "The funds need to be available before the block is delivered/executed. In the CL, this is done by getting builders to stake. In the EL, this is done with a balance check and a payment transaction." (but how can this payment transaction go through if the payload is not delivered?)

potuz replied:
The issue is mostly that the transaction can be checked if it is TOB without executing the full payload. I'll add the rewording  you indicated here, plus a short comment regarding this

barnabemonnot replied:
But what if the builder never delivers the payload? Then there is no transaction no?

potuz replied:
The way that works in the EL is that the header contains the single tx that pays, it can be executed immediately, so that the CL transition has a simple execution transition by running that single tx.

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+## Rationale
+
+### Staked Builders
```
"Staked builders"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+### Delayed validation
+
+The Payload Timeliness Committee members do not need to validate the execution payload before attesting to it. They perform basic checks like verifying the builder's signature and the correct blockhash is included. This takes away the full execution payload validation from the hot path of validation of an Ethereum block, giving the next proposer 6 seconds (`SECONDS_PER_SLOT * 2 // INTERVALS_PER_S ...
```
"basic checks such as"

barnabemonnot replied:
"verifying that the builder's signature and the correct blockhash are included"

barnabemonnot replied:
"A tradeoff from the user experience point of view is that transactions have a longer delay into being confirmed on-chain as validly included." bit confusing. Also I think it's possibly misleading to talk about "confirmation" without making explicit the confirmation rule you are referring to. Here you seem to say "the confirmation that some protocol-parties have checked validity", in which case the next proposer at `n+1` who builds on the payload at `n` would be the first party to give you this confirmation (but can you trust them?)

potuz replied:
> "verifying that the builder's signature and the correct blockhash are included"

Actually I think this is semantically different than what I say in the text. I added a comma instead. I am verifying the actual signature, not that it's included.

potuz replied:
> ...  Also I think it's possibly misleading to talk about "confirmation"...

Agreed, but the sentence is important, this is the first complain I get when describing the protocol and delayed validation. The fact that a transaction will take longer to be included in a block and longer to be in the head of the chain. I accept rephrasings... just don't know how to put it better in a single short sentence.

barnabemonnot replied:
> > ...  Also I think it's possibly misleading to talk about "confirmation"...
>
> Agreed, but the sentence is important, this is the first complain I get when describing the protocol and delayed validation. The fact that a transaction will take longer to be included in a block and longer to be in the head of the chain. I accept rephrasings... just don't know how to put it better in a single short sentence.

How about "From a user UX perspective, a transaction included in slot `n` by the builder is not widely validated until the proposer of slot `n+1` releases their beacon block on top of block `n` first, and the attesters of slot `n+1` vote on this beacon block as the head of the chain."

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+The Payload Timeliness Committee members do not need to validate the execution payload before attesting to it. They perform basic checks like verifying the builder's signature and the correct blockhash is included. This takes away the full execution payload validation from the hot path of validation of an Ethereum block, giving the next proposer 6 seconds (`SECONDS_PER_SLOT * 2 // INTERVALS_PER_S ...
+
+
+### (Block, Slot, Payload) - Forkchoice
```
Still liking "fork choice" better but I will stop noting it :)

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+- Builder reveal safety
+- Builder withhold safety. 
+
+Proposer unconditional payment refers to the following. An Ethereum slot can be either
```
"either:"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+### Reorg resistance 
+
+`PROPOSER_SCORE_BOOST` is reduced from 40 to 20. This does not allow however a en-anti reorg by a proposer with 20% of the total stake as the attacker's payload will also be not included. Proposers are protected against *1-slot sandwich* reorgs by a colluding set of validators and builders controlling more than 20% of the total stake
```
"an ex-ante"

barnabemonnot replied:
"will also not be included"

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+### Builder safety
+
+- A colluding set of proposers and attesters controlling consecutive blocks and more than 20% of the total stake can reorg a builder's payload and force it to pay the bid's value. 
+- There is no possible *unbundling* of the builder's payload, that is, if the builder reveals a payload for the head of the chain, no other payload is possible for the current slot. 
```
"No possible unbundling" it sounds like you are talking about equivocation? Unbundling = other parties can re-include stuff from the builder's block in the canonical chain, forking the builder's payload out

potuz replied:
Yes this is impossible as l well as long as the builder locks txs to the slot. On MEV boost you can actually unbundle in the same slot, on ePBS this is impossible

potuz replied:
added the words to make " There is no possible *unbundling* of the builder's payload in the same slot,..."

barnabemonnot replied:
> Yes this is impossible as l well as long as the builder locks txs to the slot. On MEV boost you can actually unbundle in the same slot, on ePBS this is impossible

I am not sure the builder can lock txs to the slot, but the tx senders can do that no?

barnabemonnot commented on EIPS/eip-epbs.md:
```diff
@@ -0,0 +1,260 @@
+
+### Malicious PTC
+
+The expected time for a malicious attacker, controlling 35% of the total stake, to have a majority control on the PTC is 205 000 years. 
```
Why choose 35%?

potuz replied:
Because to include a curve it would require me to add a pdf file to the assets. Didn't seem particularly important, should I?

barnabemonnot replied:
Ah no I think it's fine but just wondering if 35% holds some specific importance. Elsewhere you mention 20% adversary for the proposer boost, 33.33333...% is known to be a critical threshold too.

potuz replied:
Ahh right, the 20% for proposer boost is actually computed, the argument is in https://ethresear.ch/t/payload-boosts-in-epbs/18769, it's essentially the only solution to a system of inequalities that guarantees builder and proposer safety. And it does so up to that value of \beta. For the PTC ownership there's a curve, for whatever stake the attacker has, what's the expected waittime until they own the PTC?.

potuz committed:
Barnabe's review #1

potuz committed:
Barnabé's review #2

potuz committed:
Add P2P and rationale changes

Adds a P2P changes in the specification section
Adds Builder's boost on the rationale section
Adds PTC equivocations on the rationale section

potuz committed:
Barnabé review 3

potuz committed:
barnabe review 4

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+---
+eip: 9999
+title: Enshrined Proposer Builder Separation
+description: Separates the ethereum block in consensus and execution parts, adds a mechanism for the consensus proposer to chose the execution proposer.
```
Separates the Ethereum block in consensus and execution parts sounds like a structure change, but if you make it to 'Separate the **processing** of Ethereum block in consensus and execution' it is more powerful of a tag line IMO.

barnabemonnot replied:
Also chose => choose

barnabemonnot replied:
And execution proposer => builder

potuz replied:
There is a linter that forbids me to add the words " processing of the"  because it's already at the word limit. I changed chose => choose, but I am not certain about "builder" . Execution proposer only has this specific meaning that you have in mind in a niche of research on execution tickets. Here the sentence is really about consensus and execution separation and it looks better IMO to keep this symmetry instead of adding the word builder. I'll try this once, but if there's any pushback I can change that one for builder cause it passes the linter.

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+eip: 9999
+title: Enshrined Proposer Builder Separation
+description: Separates the ethereum block in consensus and execution parts, adds a mechanism for the consensus proposer to chose the execution proposer.
+author: Francesco D'Amato <francesco.damato@ethereum.org>, Barnabé Monnot <barnabe.monnot@ethereum.org>, Michael Neuder <michael.neuder@ethereum.org>, Potuz (@potuz), Terence Tsao <ttsao@offchainlabs.com>
```
Not sure if it's intended, @potuz you forgot your email

barnabemonnot replied:
don't @ him ;P

potuz replied:
Need at least one github address to be the owner of the branch

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+
+This EIP solves a different set of unrelated important problems. 
+
+- An overwhelming majority of beacon block proposers outsource the construction of the execution payload within their blocks to a third party (henceforth called a *builder*). In order to do so, they request the hash tree root (HTR) of a promised execution payload and submit a `SignedBlindedBeaconBlock` to a trusted party that is tasked with replacing the HTR with the full execution payload (recei ...
```
`hash tree root (HTR) of a promised execution payload` is fine but it's not technically correct. It's hash tree root of a promised transactions.

Same with `HTR with the full execution payload` -> `HTR with full transactions`

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+- Currently, validators have the time between receiving the full beacon block (including an execution payload) and the attesting deadline (4 seconds in Ethereum mainnet) to perform both consensus and execution state transition functions, check blob data availability and evaluate the new head of the blockchain. The remainder of the slot time is spent doing less CPU-intense and critical task. By se ...
+- By removing the full execution payload size from the consensus block, it allows for faster network propagation on the critical path. 
+- It removes the increased reorg likeliness of including blob transactions in blocks given the natural increase in timelines for DA availability checks and the fact that the builder may broadcast the blob sidecars even before attestations for the beacon block have been released. 
+- It prevents validators from missing attestations and strengthens the LMD weight properties of fork choice in the event that builders produce invalid payloads. 
```
Why `LMD weight properties of fork choice` to be specific? Why not FFG? These attestations should still count for source and target. Maybe it's cleaner to remove the LMD part or include FFG too.

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+
+## Abstract
+
+This EIP fundamentally changes the way an Ethereum block is validated by decoupling the execution validation from the consensus validation both logically as well as temporally. It does so by introducing a new optional attribution (being a *builder*) and a new duty (submitting *payload timeliness attestations*) to Ethereum validators. The `ExecutionPayload` field of the `BeaconBlockBody` is remove ...
```
Is `blockhash` supposed to be one word or two words `block hash`? We've used `blockhash` consistently across.

potuz replied:
I've used one word everywhere I hope.

terencechain commented on EIPS/eip-9999.md:
````diff
@@ -0,0 +1,285 @@
+    signature: BLSSignature
+```
+
+The `BeaconState` container is modified with the addition of:
````
We should add that `ExecutionPayloadHeader` has been modified completely. It's now the builder's header instead of the real execution block header, caching what was processed for the execution layer header since the Merge.

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+- `process_withdrawal_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- `process_consolidation_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- A new `process_payload_attestation` is added to `process_operations`, this function validates the payload timeliness attestations broadcast by the PTC members. 
+- `process_execution_payload` is now called as a separate helper when receiving a `SignedExecutionPayloadEnvelope` on the P2P layer. This function in particular checks that the HTR of the resulting beacon state coincides with the committed one in the payload envelope. 
```
I think `HTR of the resulting beacon state` can be more clearly expressed. Perhaps something like `HTR of the cached builder header in beacon state`. Right now it sounds like we are comparing the payload envelope with the beacon state.

potuz replied:
no, I really meant HTR of the resulting beacon state, there's a state transition in this function.

terencechain replied:
Ah thanks. This is my misunderstanding. I thought you meant to check the header matches the payload. Resolved

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+- `latest_block_hash`, of type `Hash32`, to track the blockhash of the last execution payload in the blockchain. 
+- `latest_full_slot`, of type `Slot`, to track the latest slot in which an execution payload was included. 
+- `latest_widthdrawals_root` of type `Root` to track the hash tree root of the latests withdrawals deducted in the beacon chain when processing a `SignedBeaconBlock`. 
+
```
Should we mention `BeaconBlockBody` is modified with the addition of aggregated payload attestations?

terencechain commented on EIPS/eip-9999.md:
```diff
@@ -0,0 +1,285 @@
+
+#### Slot auctions
+
+A simple change to this EIP is to remove the blockhash commitment from the `SignedExecutionPayloadHeader`. This allows the builder to commit any payload to the slot. A preliminary security analysis shows that payload equivocation does not weaken fork choice's FFG. Some advantages of Slot auctions include:
```
Should it be 'slot auctions' or 'Slot Auctions'? Not sure why 'slot' is capitalized.

potuz replied:
because it's a header title.

potuz committed:
Terence's review 1

potuz committed:
Barnabe review 5

potuz committed:
Add dash to title

potuz committed:
Add initial CL repo links
